Training Resources
-----------------

# Mandatory Onboarding Steps

- There are a few mandatory onboarding steps for new lab members, and they are listed on the Im Lab Onboarding Document.

# Training Resource Groups
We work with many different tools, on many different projects. So the training resources are organized into functional groups. You may want to skip reading the material in some groups, and it may be worthwhile to spend a long time with other groups. 
- [GitHub](#github)
- [Introduction to Data Science](#introduction-to-data-science)
    - [Machine Learning and Statistics](#machine-learning-and-statistics)
    - [Python](#python)
    - [R](#r)
    - [Unix](#unix)
- [Genomics](#genomics)
    - [Introduction to Genomics](#introduction-to-genomics)
    - [Lab-Specific Genomics Papers](#lab-specific-genomics-papers)
- [WorkflowR](#workflowr)
- [Computational Resources](#Computational-resources)
    - [CRI Gardner](#cri-garnder)
    - [Bionimbus PDC](#bionimbu-pdc)
    - [BigQuery](#bigquery)
- [Miscellaneous](#Miscellaneous)


# GitHub
We use GitHub to store and organize our code. There is a introduction [here](http://kbroman.org/github_tutorial/). If you are curious about when one would use certain GitHub features, look at this  [link](https://guides.github.com/introduction/flow/) which describes 'GitHub flow'.

# Introduction to Data Science
Some introductory materials give you an introduction to Git, Python, R, and machine learning topics all at once!
- Basics of Unix, Github, and Python/R [link](https://software-carpentry.org/lessons/)

## Machine Learning and Statistics
- An introduction to machine learning problems and model metrics: [link](https://rafalab.github.io/dsbook/introduction-to-machine-learning.html)
- We work fairly heavily with the generalized linear model, so it may be good to brush up on it:
    - [Generalized Linear Models](https://towardsdatascience.com/generalized-linear-models-8738ae0fb97d)
    - [Wikipedia](https://en.wikipedia.org/wiki/Generalized_linear_model)
    
## Python
## R
- Hands on 4hr intro to R (DataCamp course) [link](https://www.datacamp.com/courses/free-introduction-to-r)
- R Studio's cheatsheets: [link](https://www.rstudio.com/resources/cheatsheets/)
- Data Manipulation in R with dplyr [link](https://genomicsclass.github.io/book/pages/dplyr_tutorial.html)
- Data Visualization in R with ggplot2 [link](http://r-statistics.co/Complete-Ggplot2-Tutorial-Part1-With-R-Code.html)
- Statistical modeling in R [link](https://www.datacamp.com/courses/statistical-modeling-in-r-part-1)
- Hadley's R Style [link](http://adv-r.had.co.nz/Style.html)
## Unix

# Genomics

## Introduction to Genomics

- Intro to genomics videos [link](https://www.big-bio.org/)

## Lab-Specific Genomics Papers

# WorkflowR

# Computational Resources

## CRI Gardner
Gardner is a large high-performance computing cluster and data storage system. We use it to run computation and store data. The lab's group folder is located at `/gpfs/data/im-lab/`
- UChicago CRI Workshop Tutorials: CRI does a seminar series each academic year. Here you can find the schedule: [link](https://cri.uchicago.edu/seminar-series/)
- Intro to Gardner: this is a good explanation of what Gardner does, and why a high-performance computing cluster is important to bioinformatics: [link](http://cri.uchicago.edu/wp-content/uploads/2017/04/Gardner-Part-1.pdf)
## Bionimbus PDC
Bionimbus Protected Data Cloud is a storage/computation resource where the lab is alotted a certain amount of processors and storage, and we store and compute on virtual machines. If you'll be working on Bioinimbus, make sure to begin your application(s) quickly because the process has multiple steps.
- [Documentation](https://www.opensciencedatacloud.org/support/pdc.html) 
## BigQuery

# Miscellaneous


- Read [genomic data user code of conduct](https://osp.od.nih.gov/wp-content/uploads/Genomic_Data_User_Code_of_Conduct.pdf)
- Reproducible Research [link](http://kbroman.org/steps2rr/)

- Data Science Course [link](https://rafalab.github.io/dsbook/introduction.html)

- High Performance Computing Cluster at the UoC [link](https://rcc.uchicago.edu/resources/high-performance-computing)
	- RCC training schedule [link](https://rcc.uchicago.edu/support-and-services/workshops-and-training)
	- Request account [link](https://rcc.uchicago.edu/accounts-allocations/request-account)
	
- Get CITI training [link](https://www.citiprogram.org)
	- Basics of Health Privacy
	- Responsible Conduct of Research (RCR) Basic
	- Human Subjects Research – Biomedical
	- Basics of Information Security
	- Conflict of Interest




### Recommended Training Material



- Coursera courses (these can be taken for free)
	- Introduction to Genomic Technologies [link](https://www.coursera.org/course/genintro)
	- Genomic Data Science with Galaxy
	- Python for Genomic Data Science [link](https://www.coursera.org/course/genpython)
	- Algorithms for DNA Sequencing
	- Command Line Tools for Genomic Data Science
	- Bioconductor for Genomic Data Science
	- Statistics for Genomic Data Science [link](https://www.coursera.org/course/genstats) 



- BigQuery tutorials [link](https://cloud.google.com/bigquery/docs/tutorials)

- Google Cloud training document [link](https://docs.google.com/document/d/1z35R9uZ2iDo-Fp1ImUMBIKvcUunBTUp1ZmAXZF2Hf48/edit?usp=sharing)

### Suggested Training Material

- The New Genetics [link](http://publications.nigms.nih.gov/thenewgenetics/index.html)

- Data Science Track Coursera 
	- The Data Scientist’s Toolbox [link](https://www.coursera.org/course/datascitoolbox)
	- R Programming [link](https://www.coursera.org/course/rprog)
	- Getting and Cleaning Data [link](https://www.coursera.org/course/getdata)
	- Exploratory Data Analysis [link](https://www.coursera.org/course/exdata)
	- Practical Machine Learning [link](https://www.coursera.org/course/predmachlearn)
	- Statistical Inference [link](https://www.coursera.org/course/statinference)
	- Reproducible Research [link](https://www.coursera.org/course/repdata)

### Recommended Reading

Gamazon ER, Wheeler HE, Shah KP, Mozaffari SV, Aquino-Michaels K, Carroll RJ, Eyler AE, Denny JC, Nicolae DL, Cox NJ, Im HK: A gene-based association method for mapping traits using reference transcriptome data. Nat Genet 2015:1–10.

GTEx Consortium: The Genotype-Tissue Expression (GTEx) pilot analysis: multitissue gene regulation in humans. Science 2015, 348:648–660.

The 1000 Genomes Consortium: A global reference for human genetic variation [link](http://www.nature.com/doifinder/10.1038/nature15393)

Kundaje A, Meuleman W, Ernst J, Bilenky M, Yen A, Heravi-Moussavi A, et al. Integrative analysis of 111 reference human epigenomes. Nature. 2015;518:317–30. 

Li YI, van de Geijn B, Raj A, Knowles DA, Petti AA, Golan D, et al. RNA splicing is a primary link between genetic variation and disease. Science. American Association for the Advancement of Science; 2016;352:600–4. 

Albert FW, Kruglyak L: The role of regulatory variation in complex traits and disease. Nat Rev Genet 2015, 16:197–212.

Das S, Abecasis GR, Browning BL: Genotype Imputation from Large Reference Panels. Annu Rev Genomics Hum Genet 2018;19:73-96.

Im HK, Gamazon ER, Nicolae DL, Cox NJ: On sharing quantitative trait GWAS results in an era of multiple-omics data and the limits of genomic privacy. Am J Hum Genet 2012, 90:591–598.

Finucane HK, Bulik-Sullivan B, Gusev A, Trynka G, Reshef Y, Loh P.-R., et al. Partitioning heritability by functional annotation using genome-wide association summary statistics. Nature Genetics 2015, 47:1228-1235.

Finucane HK, Reshef YA, Anttila V, Slowikowski K, Gusev A, Byrnes A, et al. Heritability enrichment of specifically expressed genes identifies disease-relevant tissues and cell types. Nature Genetics 2018, 50:621-629.

Visscher PM: Human Complex Trait Genetics in the 21st Century. Genetics 2016, 202:377–379.

### Additional R Tutorials
- [dplyr](https://cran.rstudio.com/web/packages/dplyr/vignettes/introduction.html)
- [RHadoop](https://github.com/andrie/RHadoop-tutorial/)
- [Rocker](http://dirk.eddelbuettel.com/papers/useR2015_docker.pdf)
- [Machine learning mlr](https://mlr.mlr-org.com/index.html)
- [workflowr](https://jdblischak.github.io/workflowr/articles/wflow-01-getting-started.html)
- [Enloc-coloc comparison](https://hakyimlab.github.io/gtex-gwas-hki/enloc-coloc-comparison-2.html)
- More DataCamp courses
	- Introduction to the Tidyverse [link](https://www.datacamp.com/courses/introduction-to-the-tidyverse)
	- Importing data in R [link](https://www.datacamp.com/courses/importing-data-in-r-part-1)
	- Cleaning data in R [link](https://www.datacamp.com/courses/cleaning-data-in-r)
	- Modeling with data in the Tidyverse [link](https://www.datacamp.com/courses/modeling-with-data-in-the-tidyverse)
	- Introduction to bioconductor [link](https://www.datacamp.com/courses/introduction-to-bioconductor)
	- Single-cell RNA-seq workflows in R [link](https://www.datacamp.com/courses/single-cell-rna-seq-workflows-in-r)


### Useful links

- [Karl Broman's Tutorials](http://kbroman.org/pages/tutorials.html)
- [Jeff Leek's Github page](https://github.com/jtleek)

### Cheatsheets


	- Data Wrangling [download](https://www.rstudio.com/wp-content/uploads/2015/02/data-wrangling-cheatsheet.pdf)
	- R Markdown [download](https://www.rstudio.com/wp-content/uploads/2015/02/rmarkdown-cheatsheet.pdf)
  	- Data visualization [download](https://www.rstudio.com/wp-content/uploads/2015/11/ggplot2-cheatsheet.pdf)

<!--### Account setup-->

<!--- Tarbell-->
<!--- Github-->
<!--- Open Science Data Cloud-->
<!--- Slack-->
<!--- AWS-->
<!--- Google Cloud-->

<!--### Contractors-->


<!--Open payoneer (if no US account available)-->

<!--Fill out forms -->

<!--- [US Individuals](https://www.dropbox.com/sh/2h3xmmxkn2i1tvy/AAADYYRsHwtO8JPJ37_dHKUTa?dl=0)-->
<!--- [Foreign Individuals](https://www.dropbox.com/sh/8eb5c21kc8ezlbn/AACe7caz-p0vbhBplzA2I4PCa?dl=0)-->

<!--### Volunteers-->

<!--Forms [link](https://www.dropbox.com/sh/k1qakuofx6pbi91/AAC-e0gnNImYTUlHqGryyXjJa?dl=0)-->
